src/Microsoft.VisualStudio.Threading/AsyncSemaphore.cs

```diff
@@ -97,7 +97,14 @@ protected virtual void Dispose(bool disposing)
         {
             if (disposing)
             {
-                this.semaphore.Dispose();
+                // !!DO NOT!! dispose the semaphore, for the following reasons:
+                // 1. SemaphoreSlim.Dispose is not safe for concurrent use while something is waiting. Calling Dispose
+                //    requires tracking of all outstanding requests for enter/release.
+                // 2. SemaphoreSlim only allocates resources that could need to be be disposed if the
+                //    SemaphoreSlim.AvailableWaitHandle property is accessed. The semaphore field here is private and we
+                //    do not access this property.
```

Comment:
@AArnott The code is resilient to `ObjectDisposedException`. The problem is that unlike the other methods on `SemaphoreSlim`, `Dispose()` isn't safe for concurrent use. In asynchronous code, this concurrency applies not just to the direct calls to `SemaphoreSlim`, but to the entire life of the task returned by `SemaphoreSlim.WaitAsync`. The situation is complicated by the fact that it is *not* sufficient to cancel the `CancellationToken` passed to `SemaphoreSlim.WaitAsync`. In order to safely call `Dispose`, you must:

1. Cancel the `CancellationToken` associated with all calls to `SemaphoreSlim.WaitAsync`
2. Wait for all tasks returned from `WaitAsync` to complete in the cancelled state

Considering `SemaphoreSlim.Dispose` is nothing more than a state invalidation when used with `AsyncSemaphore`, there is no advantage to calling it. The garbage collector exists for collecting the semaphore's managed memory when it is no longer in use. While I would not make a claim that either Roslyn or StreamJsonRpc is necessarily doing everything correctly, I would claim that there is no functional downside to including this change. As a bonus it happens to also be the shortest and least risky path to fixing a customer-visible bug in Visual Studio. It's important that it be included.